bbn-test-5: Add doc for FP v2.0.0-rc.0 upgrade (#503)
babylonlabs-io/networks: bbn-test-5/finality-provider/upgrades/v2rc0/README.md

  [here](../../../babylon-node/upgrades/v3/README.md#3-1-preparation).

Once the above are completed, wait for the Babylon Genesis Testnet
to reach the upgrade height `X` (TODO: Specify the height)
on which the node will halt for the upgrade to be applied.

### 3.2. Execution

The upgraded Babylon Genesis Finality Provider and EOTS daemon binaries are
tightly coupled with the Babylon Genesis Testnet software upgrade.

The following steps must be executed **with precise ordering and timing**
to ensure that your finality provider does not experience downtime and/or
submits invalid finality signatures:
1. **Stop the Finality Provider and EOTS daemons** once the Babylon Genesis
   Testnet upgrade height `X` (TODO: Specify the height) is reached.
2. **Swap the Babylon Genesis finality provider and EOTS daemon binaries with
   the new ones**.
3. In your Babylon Genesis Finality Provider config, append the following
   configuration to the `[Application Options]` section
   (TODO: Specify the height):
   ```shell
   [Application Options]

   ; The height at which the context signing will start
   ContextSigningHeight = X
   ```
4. **Start the Finality Provider and the EOTS Daemons.**

> 5. **Upgrade your Babylon node to `v3.0.0-rc.0` following the instructions

Author: After sync with @maiquanghiep , we have decided to move the babylon node upgrade **after** the FP upgrade.

This averts the risk of the FP receiving new Babylon blocks prior to being upgraded, and thus submitting signatures for them without context.

# `v2.0.0-rc.0` Software Upgrade

This document summarizes the procedure to upgrade a Babylon Finality Provider
to version v2.0.0-rc.0.

## Table of Contents

1. [High-level Overview](#1-high-level-overview)
2. [Technical Deep-dive](#2-technical-deep-dive)
3. [Applying the upgrade](#3-applying-the-upgrade)
   1. [Preparation](#31-preparation)
   2. [Execution](#32-execution)
   3. [Verification](#33-verification)

## 1. High-level Overview

- **Upgrade version**: `v2.0.0-rc.0`
- **Upgrade Babylon height**: `X` (TODO: specify the height)
- **Upgrade overview**: This upgrade introduces a major refactoring that
  transforms the Finality Provider into a reusable SDK, to enable the
  development of Finality Provider software for BSNs.

> - **Finality Providers who should participate**: Only Babylon Finality Providers.

Author: Won't Babylon still be responsible for the overall development of the BSN FP software?
What if we want to rollout e.g., a breaking upgrade that affects all FP software?

> - **Finality Providers who should participate**: Only Babylon Finality Providers.

Author: Synced offline, this will be taken care of on the release notes of the BSN FPs, which will live on separate repos.

## 2. Technical Deep-dive

The Babylon Genesis Testnet `v3.0.0-rc.0` upgrade (TODO: link to corresponding
upgrade doc) is coupled with a coordinated breaking upgrade of the Finality
Provider software from `v1.x.x` (`v1.0.0`, `v1.1.0-rc.0`, `v1.1.0-rc.1`) to
`v2.0.0-rc.0`.

Finality Provider `v2.0.0-rc.0` introduces a major refactoring of the codebase
into a reusable SDK. This aims to enable the development of Finality
Provider software for Bitcoin Supercharged Networks (BSNs). The following BSN
stacks are initially supported:
- Ethereum L2 Rollup BSNs, integrating via the deployment of a CosmWasm Smart
  Contract on the Babylon Genesis Testnet. The Rollup BSN Finality Provider
  will be responsible for the submission of finality signatures and public
  randomness to the Contract.
- Cosmos SDK BSNs, integrating via the deployment of a CosmWasm Smart Contract
  on their respective chains and the establishment of IBC Channels with Babylon.
  The Cosmos BSN Finality Provider will be responsible for the submission of

>   finality signature and public randomness to the Contract.

Reviewer: It's unclear whether the above is directly relevant to Babylon Genesis finality providers, especially since they won't be using the BSN software. It might be more effective to include a concise bullet list (similar to what I suggested earlier for the previous sectoin) that outlines our key focus areas. This should also emphasize our efforts around compatibility and, importantly, the signing context.

Given our concerns around the signing context, it would make sense to concentrate on technically expanding its functionality. We could also highlight potential pitfalls that implementers might face, so they’re well-prepared in advance.

binaries and operational guides. Babylon Finality Providers will continue
operating the original software.**

*BSN Finality Providers will not participate in this software upgrade, as they
will start operating their software only after the Babylon Genesis Testnet
is upgraded to the `v3.0.0-rc.0` version.*

## 3. Applying the upgrade

### 3.1. Preparation

The following actions must be taken prior to the upgrade time:
- Ensure that your Babylon Finality Provider and EOTS daemons are currently
  operating version `v1.0.0`, `v1.1.0-rc.0` or `v1.1.0-rc.1`.
- Obtain the `v2.0.0-rc.0` binary. You can achieve this in multiple ways:
  - Download the binary from the [releases
    page](https://github.com/babylonlabs-io/finality-provider/releases/tag/v2.0.0-rc.0)
  - Build the binary on your machine
    ```shell
    git checkout v2.0.0-rc.0
    make install
    ```
  - If you’re working with Docker images, you can pull the pre-built Docker image:
    ```shell
    docker pull babylonlabs/finality-provider:v2.0.0-rc.0
    ```
- Prepare for the Babylon `v3.0.0-rc.0` node upgrade following the instructions
  here (TODO: Point to the Babylon node upgrade guide)

> Once the above are completed, patiently wait for the Babylon Genesis Testnet

Reviewer: ```suggestion
Once the above are completed, wait for the Babylon Genesis Testnet
```

> to reach the upgrade height `X` (TODO: Specify the height).

Reviewer: ```suggestion
to reach the upgrade height `X` (TODO: Specify the height)
on which the node will halt for the upgrade to be applied.
```

### 3.2. Execution

> The Babylon Finality Provider and EOTS daemon upgrades are tightly coupled
> with the Babylon network software upgrade. The following steps must be executed
> with precise ordering and timing, to ensure minimum downtime and no slashing
> risk for your Finality Provider:
> - Once the Babylon upgrade height `X` (TODO: Specify the height) is reached,
>   stop your Babylon Finality Provider and EOTS daemons.
> - Upgrade your Babylon node to `v3.0.0-rc.0` following the instructions here
>   (TODO: Point to the Babylon node upgrade guide)
> - Swap the Babylon Finality Provider and EOTS daemon binaries with the latest
>   ones.
> - In your Babylon Finality Provider config, append the following configuration
>   to the `[Application Options]` section (TODO: Specify the height):
>   ```shell
>     ; The height at which the context signing will start
>     ContextSigningHeight = X
>   ```
> - Start the EOTS daemon. **DO NOT START THE BABYLON FINALITY PROVIDER DAEMON YET.**
> - Wait until the Babylon produces block `X + 1`. This can take up to 5 minutes,
>   depending on how fast Babylon CometBFT Validators upgrade their Babylon nodes.
> - Once the block is produced, start your Babylon Finality Provider daemon.

Reviewer: ```suggestion
The upgraded Babylon Genesis finality provider and EOTS daemon binaries
are tightly coupled with the Babylon Genesis Testnet software upgrade.
The following steps must be executed **with precise ordering and timing**
to ensure that your finality provider does not experience downtime and/or
submits invalid finality signatures:

1. **Stop the Finality Provider and EOTS daemons** once the Babylon Genesis
   testnet upgrade height `X` (TODO: Specify the height) is reached.
2. **Upgrade your Babylon node to `v3.0.0-rc.0` following the instructions here
  (TODO: Point to the Babylon node upgrade guide)
3. **Swap the Babylon Genesis finality provider and EOTS daemon binaries with
     the new ones**.
 4. In your Babylon Genesis Finality Provider config, append the following configuration
     to the `[Application Options]` section (TODO: Specify the height):
     ```shell
        ; The height at which the context signing will start
        ContextSigningHeight = X
      ```
5. **Start the EOTS daemon** **DO NOT START THE BABYLON FINALITY PROVIDER DAEMON YET.**
6. Wait until the Babylon produces block `X + 1`. This can take up to 5 minutes,
    depending on how fast Babylon CometBFT Validators upgrade their Babylon nodes.
7. Once the block is produced, start your Babylon Finality Provider daemon.
```

Small edit to make this a numbered list and some wording. I don't fully understand steps 5-7. Why should someone wait for a block to be produced to start their finality provider daemon?

Author: They actually do not need to wait, I've removed that part.

# `v2.0.0-rc.0` Software Upgrade

This document summarizes the procedure to upgrade a Babylon Genesis Finality
Provider to version v2.0.0-rc.0.

## Table of Contents

1. [Overview](#1-overview)
2. [Technical Deep-dive](#2-technical-deep-dive)
3. [Applying the upgrade](#3-applying-the-upgrade)
   1. [Preparation](#31-preparation)
   2. [Execution](#32-execution)
   3. [Verification](#33-verification)

## 1. Overview

- **Upgrade version**: `v2.0.0-rc.0`
- **Upgrade Babylon height**: `X` (TODO: specify the height)
- **Upgrade overview**:
  - **Compatibility with Babylon Genesis v3**
  - **Usage of a signing context** to combat the usage of existing signatures for
    separate contexts.
  - **Major refactoring** of the finality provider codebase that
    transforms the Finality Provider into a reusable SDK, to enable the
    development of Finality Provider software for BSNs.

## 2. Technical Deep-dive

> The Babylon Genesis Testnet `v3.0.0-rc.0` upgrade (TODO: link to corresponding

Reviewer: Wouldn't this be `../../../babylon-node/upgrades/v3/README.md`?

    binaries and operational guides. Babylon Finality Providers will continue
    operating the original software.

## 3. Applying the upgrade

### 3.1. Preparation

The following actions must be taken prior to the upgrade time:
- Ensure that your Babylon Finality Provider and EOTS daemons are currently
  operating version `v1.0.0`, `v1.1.0-rc.0` or `v1.1.0-rc.1`.
- Obtain the `v2.0.0-rc.0` binary. You can achieve this in multiple ways:
  - Download the binary from the [releases
    page](https://github.com/babylonlabs-io/finality-provider/releases/tag/v2.0.0-rc.0).
  - Build the binary on your machine:
    ```shell
    git checkout v2.0.0-rc.0
    make install
    ```
  - Pull the pre-built Docker image:
    ```shell
    docker pull babylonlabs/finality-provider:v2.0.0-rc.0
    ```
- Prepare for the Babylon `v3.0.0-rc.0` node upgrade following the instructions
  here (TODO: Point to the Babylon node upgrade guide).

Once the above are completed, wait for the Babylon Genesis Testnet
to reach the upgrade height `X` (TODO: Specify the height)
on which the node will halt for the upgrade to be applied.

### 3.2. Execution

> The Babylon Finality Provider and EOTS daemoThe upgraded Babylon Genesis

Reviewer: ```suggestion
The upgraded Babylon Genesis
```

Finality Provider and EOTS daemon binaries are tightly coupled with the Babylon
Genesis Testnet software upgrade.

The following steps must be executed **with precise ordering and timing**
to ensure that your finality provider does not experience downtime and/or
submits invalid finality signatures:
1. **Stop the Finality Provider and EOTS daemons** once the Babylon Genesis

>    Testnet upgrade height `X` (TODO: Specify the height) is reached.

Reviewer: @KonradStaniec @Lazar955 if this happens after the height is reached (even a second), we risk a signature without a context being sent by the fp and included in the mempool.

Author: A signature without context will be sent for block `X - 1` when the chain halts. This signature will have to be processed after the chain is upgraded. Will the node actually be able to process it? @KonradStaniec @Lazar955 

After the halt, block `X` will be produced, and the FP will submit a sig for block `X` with signing context.

Reviewer 2: hmm I think we need to play it step by step so everybody is aligned. 

First, I would leave the consensus block out of the picture as it is not relevant to FP operations. The only thing that matters is execution block.

Second, important thing is that we all know all blocks are indexed in begin blocker - https://github.com/babylonlabs-io/babylon/blob/v3.0.0-snapshot.250731/x/btcstaking/keeper/keeper.go#L114.

So given all that and the fact that upgrade happens at block `X`:

1. block `X-1` is executed and indexed in the `BeginBlocker`
2. Now fps will start voting on it **without** context and those transactions will end up in the mempool and ultimately into consensus committed block `X`
3. block `X` is executed with old binary, it contains all votes for block `X-1`. It fails to be committed, as it is halt height. 
4. binary is swapped
5. block `X` is executed with new binary. 

Hmm from this it seems that we should set `ContextSigningHeight` to `UpgradeHeight -1 ` so that new binary started at height `X` can verify votes for block `X-1`

Reviewer 2: >. If somehow the FP vote for block X - 1 gets lost from mempool during babylon node restarts,

when Babylon node restarts due to upgrade, the transactions will already be in the block i.e block `X` will be already decided by consensus.

2. **Upgrade your Babylon node to `v3.0.0-rc.0` following the instructions here.
   (TODO: Point to the Babylon node upgrade guide)
3. **Swap the Babylon Genesis finality provider and EOTS daemon binaries with
   the new ones**.
4. In your Babylon Genesis Finality Provider config, append the following
   configuration to the `[Application Options]` section
   (TODO: Specify the height):
   ```shell
   [Application Options]

   ; The height at which the context signing will start

>    ContextSigningHeight = X + 1

Reviewer: @KonradStaniec @Lazar955 `X` or `X+1`?

Reviewer 3: `X`

   ```
5. **Start the Finality Provider and the EOTS Daemons.**

> 6. Wait until the Babylon produces block `X + 1`. This can take up to 5 minutes,

Reviewer 3: X instead of X+1

  - Query the Babylon node directly, replacing `FP_BTC_PK_HEX` with the BTC
    public key of your Babylon Finality Provider in hex format. The resulting
    height should be equal or greater to `X + 1` (TODO: Specify the height).
    ```shell
    babylond q btcstaking finality-provider \
      FP_BTC_PK_HEX \
      --node https://babylon-testnet-rpc.polkachu.com:443 -o json \
      | jq -r .finality_provider.highest_voted_height
    ```
  - Check your Babylon Finality Provider logs, and look for a log of the
    following format:
    ```shell
    LOG_TIMESTAMP	info	successfully submitted the finality signature to the consumer chain	{"consumer_id": "bbn-test-5", "pk": "FP_BTC_PK_HEX", "start_height": X, "end_height": X, "tx_hash": "TX_HASH"}
    ```
  - A network explorer can also be consulted (examples:
   [Xangle](https://babylon-explorer.xangle.io/testnet/finality-providers),
   [Nodes.guru](https://testnet.babylon.explorers.guru/finality-providers),
   [Mintscan](https://www.mintscan.io/babylon-testnet/finality-providers)).
   Explorers have a data indexing overhead, so it's likely that your finality
   signatures will be reflected after ~1 minute.

> - Verify that your Babylon Finality Provider is not in `Slashed` state. You

Reviewer: This check is a bit weird, an fp shouldn't have to worry about getting slashed in an upgrade

  - Thus, it's critical that the Babylon Genesis Finality Providers upgrade
    their software and properly configure the context signing height no later
    than the Babylon Genesis Testnet upgrade height `1692200`.
    **For safety, Finality Provider operators must upgrade
    their software earlier than the Babylon Genesis Testnet network upgrade.**
- Major refactoring of the codebase into a reusable SDK. This aims to enable
  the development of Finality Provider software for Bitcoin Supercharged
  Networks (BSNs). The initial BSN stacks will include Ethereum L2 Rollups and
  Cosmos SDK chains.
  - It's important to emphasize that the BSN Finality Provider software will be
    distinct from the Babylon Finality Provider software, coming with dedicated
    binaries and operational guides. Babylon Finality Providers will continue
    operating the original software.

## 3. Applying the upgrade

Applying the upgrade constitutes of 2 steps:
- Upgrade your Finality Provider and EOTS daemons **before** the Babylon Genesis
  Testnet upgrade height `1692200` is reached.
- Upgrade your Babylon node once the network halts.

### 3.1. Upgrade the Finality Provider

> **NOTE: THIS MUST HAPPEN BEFORE THE BABYLON GENESIS TESTNET UPGRADE HEIGHT

Reviewer: ```suggestion
> **⚠️ NOTE: THIS MUST HAPPEN BEFORE THE BABYLON GENESIS TESTNET UPGRADE HEIGHT
```

  operating version `v1.0.0`, `v1.1.0-rc.0` or `v1.1.0-rc.1`.
- Obtain the `v2.0.0-rc.0` binary. You can achieve this in multiple ways:
  - Download the binary from the [releases
    page](https://github.com/babylonlabs-io/finality-provider/releases/tag/v2.0.0-rc.0).
  - Build the binary on your machine:
    ```shell
    git checkout v2.0.0-rc.0
    make install
    ```
  - Pull the pre-built Docker image:
    ```shell
    docker pull babylonlabs/finality-provider:v2.0.0-rc.0
    ```

#### 3.1.2. Execution

The upgraded Babylon Genesis Finality Provider and EOTS daemon binaries are
tightly coupled with the Babylon Genesis Testnet software upgrade.

The following steps must be executed **before the Babylon Genesis Testnet
upgrade height is reached**, to ensure that your finality provider does not
experience downtime or submit invalid signatures:
1. **Stop the Finality Provider and EOTS daemons**.
2. **Swap the Babylon Genesis finality provider and EOTS daemon binaries with
   the new ones**.
3. In your Babylon Genesis Finality Provider config, append the following
   configuration to the `[Application Options]` section:
   ```shell
   [Application Options]

> 

Reviewer 4: Found some other updates in the config as well
```
; The upper bound of the number of Schnorr public randomness for each commitment
NumPubRandMax = 500000

; The interval between each attempt to commit public randomness
RandomnessCommitInterval = 30s
```

Also this config is removed in v2
```
; the type of the consumer chain
ChainType = babylon
```

Verify that your Babylon Finality Provider is voting as expected:
  - Query the Babylon node directly, replacing `FP_BTC_PK_HEX` with the BTC
    public key of your Babylon Finality Provider in hex format. The height
    should be growing.
    ```shell
    babylond q btcstaking finality-provider \
      FP_BTC_PK_HEX \
      --node https://babylon-testnet-rpc.polkachu.com:443 -o json \
      | jq -r .finality_provider.highest_voted_height
    ```
  - Check your Babylon Finality Provider logs, and look for logs of the
    following format:
    ```shell
    LOG_TIMESTAMP	info	successfully submitted the finality signature to the consumer chain	{"consumer_id": "bbn-test-5", "pk": "FP_BTC_PK_HEX", "start_height": X, "end_height": X, "tx_hash": "TX_HASH"}
    ```
  - A network explorer can also be consulted (examples:
   [Xangle](https://babylon-explorer.xangle.io/testnet/finality-providers),
   [Nodes.guru](https://testnet.babylon.explorers.guru/finality-providers),
   [Mintscan](https://www.mintscan.io/babylon-testnet/finality-providers)).
   Explorers have a data indexing overhead, so it's likely that your finality
   signatures will be reflected after ~1 minute.

> ### 3.2. Upgrade the Babylon node

Reviewer: ```suggestion
### 3.2. Upgrade the Babylon Genesis node
```

why not just link to the upgrade guide we'll have for the node?

#### 3.2.1. Preparation

Prepare for the Babylon `v3.0.0-rc.0` node upgrade following the instructions
[here](../../../babylon-node/upgrades/v3/README.md#3-1-preparation).

#### 3.2.2. Babylon Node Upgrade

Once the Babylon Genesis Testnet upgrade height `1692200` is reached, perform the
following steps:
1. **Upgrade your Babylon node to `v3.0.0-rc.0` following the instructions
   [here](../../../babylon-node/upgrades/v3/README.md#3-2-execution).**
2. Wait until the Babylon produces block `1692200`. This can take up to 5 minutes,
   depending on how fast Babylon CometBFT Validators upgrade their Babylon
   nodes.

#### 3.2.3. Verification

> After completing the [execution](#322-execution) section, perform the following

Reviewer 4: `After completing the [execution](#322-execution) section of the Finality Provider upgrade, perform the following`
